File: src/main/scala/edu/uchicago/cs/encsel/ndnn/Graph.scala
/**
 * *****************************************************************************
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 * Contributors:
 *     Hao Jiang - initial API and implementation
 *
 * *****************************************************************************
 */
package edu.uchicago.cs.encsel.ndnn

import scala.collection.GenIterable
import scala.collection.mutable.ArrayBuffer

import org.nd4j.linalg.api.ndarray.INDArray

class Graph(ip: InitPolicy, up: UpdatePolicy, loss: LossFunction) extends NodeEnv {

  val initPolicy = ip
  val updatePolicy = up
  val lossFunction = loss

  protected val inputs = new ArrayBuffer[Input]
  protected val params = new ArrayBuffer[Param]
  protected val expected = new Input(this)
  protected var out: Node = null

  def param(n: String, shape: Array[Int])(implicit usePolicy: InitPolicy = initPolicy): Param = {
    val newparam = new Param(n, this)
    newparam.value = usePolicy.init(shape)
    params += newparam
    newparam
  }

  def input(n: String, src: Node*): Input = {
    val newinput = new Input(n, this, src: _*)
    inputs += newinput
    newinput
  }

  def expect(value: INDArray) = expected.setValue(value)
  def output(node: Node): Unit = out = node

  def getInputs = inputs.clone()
  def getParams = params.clone()
  def getOutput = out

  def train: Double = {
    forward
    // Compute Loss
    val loss = lossFunction.loss(out.value, expected.value)

    // Backward
    out.grad = lossFunction.gradient
    backward
    // Update Parameters and Decay Weight
    params.foreach { updatePolicy.update(_) }
    updatePolicy.weightDecay()
    loss
  }

  def test: (Double, Int) = {
    forward
    // Compute Loss
    if (out != null && lossFunction != null) {
      val loss = lossFunction.loss(out.value, expected.value, true)
      (loss, lossFunction.accuracy)
    } else {
      (-1, -1)
    }
  }

  def dump(): Array[INDArray] = this.params.map { _.value }.toArray
  def load(params: GenIterable[INDArray]) = this.params.zip(params).foreach(p => p._1.value = p._2)
}

File: src/main/python/ndnn/graph.py
from ndnn.node import Param, Input
from ndnn.init import Xavier

class Graph(object):
    def __init__(self, loss, update):
        self.params = []
        self.inputs = []
        self.out = None
        self.loss = loss
        self.update = update
    
    def input(self):
        x = Input()
        self.inputs.append(x)
        return x
    
    def param(self):
        param = Param()
        self.params.append(param)
        return param
    
    def param_of(self, shape, init=Xavier()):
        param = Param()
        param.value = init.apply(shape)
        self.params.append(param)
        return param
        
    def output(self, node):
        self.out = node

    def expect(self, value):
        self.expect_val = value

    def train(self):
        for x in self.inputs:
            x.forward(x)
        for p in self.params:
            p.forward(p)
        # Compute loss
        loss_val = self.loss.loss(self.out.value, self.expect_val, False)
        # Backward
        self.out.backward(self.out, self.loss.grad)
        for p in self.params:
            self.update.update(p) 
        return loss_val, self.loss.accuracy()
    
    def test(self):
        for x in self.inputs:
            x.forward(x)
        for p in self.params:
            p.forward(p)
        # Compute loss
        if self.loss is not None:
            loss_val = self.loss.loss(self.out.value, self.expect_val, True)
            return loss_val, self.loss.accuracy()
        else:
            return -1., -1
    
    def dump(self):
        return [p.value for p in self.params]
    
    def load(self, ps):
        for i, p in enumerate(ps):
            self.params[i].value = p
